File: design_patterns/c++/src/base/Room.cpp
#include "../../headers/Room.h"
#include "../../headers/MapSite.h"

Room::Room(int roomNo) {
  _roomNumber = roomNo;
}

Room::Room(const Room & other){
  _roomNumber = other._roomNumber;
  for (int i = 0; i < MAP_SIDE_SIZE; ++i) {
	_side[i] = dynamic_cast<MapSite*>(other._side[i]->clone());
  }
}

MapSite* Room::getSide(Direction direction) const {
  return _side[direction];
}

void Room::setSide(Direction direction, MapSite* mapSite) {
  _side[direction] = mapSite;
}

void Room::setNum(int num) {
  _roomNumber = num;
}

Room* Room::clone() const {
  return new Room(*this);
}

File: design_patterns/java/src/base/Room.java
package base;

public class Room implements MapSite {
    int mRoomNo;
    MapSite[] mSide;
    public Room(int num) {
        mRoomNo = num;
        mSide = new MapSite[4];
    }

    public boolean enter() {
        return true;
    }

    public void setSide(Direction direction, MapSite site) {
        switch(direction) {
            case East:
                mSide[0] = site;
                break;
            case South:
                mSide[1] = site;
                break;
            case West:
                mSide[2] = site;
                break;
            case North:
                mSide[3] = site;
                break;
        }
    }

    public MapSite getSide(Direction direction) {
        switch(direction) {
            case East:
                return mSide[0];
            case South:
                return mSide[1];
            case West:
                return mSide[2];
            case North:
                return mSide[3];
            default:
                return null;
        }
    }

    @Override
    public Room clone() {
        return new Room(this.mRoomNo);
    }
}
